### PR TITLE
implements vim mode

### LINKED ISSUE #1982: Feature Request: Vim Mode
#### Nature of issue?

- New feature request

#### New feature details:

This feature request together with a working implementation proposes a new user preference to enable Vim Mode on p5 Codemirror editor.

I did some search about any discussion related to Vim on p5 web editor, but I couldn't find any, so I ended up with a working implementation which wasn't that difficult as I thought, since Codemirror already supports Vim Keymap, I just had to change the Keymap from `codemirror/keymap/sublime` to `codemirror/keymap/vim` and add a preference option to the UI.

Screen Shot 2021-11-26 at 02 03 33

One small concern is the slightly increase on the bundle size, since `codemirror/keymap/vim` is about 220 KB, (although I ended up with 4.2MB on my app.[hash].js on my local build and 4.1MB without `keymap/vim`, but on the live p5 editor is just 1.1MB, I am a bit confused 🧐).

Another concern to keep in mind is with underlying issues with Codemirror Vim Keymap implementation itself, which apparently does not have an [active maintainer](https://codemirror.net/demo/vim.html), although I didn't had any problems.

I am not sure if this is a such demanded feature, but would love to hear some feedback.

Code: https://github.com/cesardeazevedo/p5.js-web-editor/commit/03fe75b3d60c5f52375758993e565441929ebfa7
Implementation: https://p5-vim.herokuapp.com/ (I don't know how to fix the `PREVIEW_URL`)
Codemirror vim issues: https://github.com/codemirror/codemirror/issues?q=is%3Aissue+is%3Aopen+label%3Avim

### PR DESCRIPTION
**Pull request by juscuzryancan**
Fixes #1982 

Changes:

I have verified that this pull request:

* [x] has no linting errors (`npm run lint`)
* [x] has no test errors (`npm run test`)
* [x] is from a uniquely-named feature branch and is up to date with the  `develop` branch.
* [x] is descriptively named and links to an issue number, i.e. `Fixes #123`

### CODE REVIEW

File: client/modules/IDE/components/Editor.jsx [hunk 1]
```diff
@@ -264,6 +265,9 @@ class Editor extends React.Component {
         CodeMirror.showHint(this._cm, () => {}, {});
       }
     }
+    if (this.props.vimMode !== prevProps.vimMode) {
+      this._cm.setOption('vimMode', this.props.vimMode);
```
Comment (lindapaiste) [inline client/modules/IDE/components/Editor.jsx]: The option that we are changing on the CodeMirror instance is the `keyMap`.  Untested, but I think it should be this.  Basically just repeating how you set it up initially but based on the new `vimMode` setting.

```suggestion
      this._cm.setOption('keyMap', this.props.vimMode ? 'vim' : 'sublime');
```

Comment (juscuzryancan) [reply client/modules/IDE/components/Editor.jsx]: Thanks for pointing that out!

**Comment by lindapaiste** [comment]
If the user has enabled the Vim keymap, then there are a bunch of shortcuts that we are displaying in the UI would which be inaccurate, right?  I don't know which of these are specific to the current `sublime` keymap but we have a bunch of shortcuts displayed in the [`KeyboardShortcutModal`](https://github.com/processing/p5.js-web-editor/blob/develop/client/modules/IDE/components/KeyboardShortcutModal.jsx) and in various nav menus.  If a lot of those are impacted then maybe we need some sort of centralized utility for looking up the current command so that we don't have to have a million ternaries everywhere.

I'm also thinking about how we would display this in the preferences because I'm concerned that "Vim Mode: On / Off" is not intuitive if the user doesn't know what Vim is.  Maybe some sort of explanatory tooltip would suffice.  I'm pondering if "Keyboard Shortcuts: Sublime / Vim" is clearer.  My concern is that someone might inadvertently turn Vim Mode on and then be really confused as to why the keyboard shortcuts that they are used to don't do anything.

**Comment by raclim** [comment]
Since it's been a while since this PR has been updated, I've marked it with a "closing soon" label, which signifies that this PR will be closed in 10 days. Please feel free to revisit this to make any further updates or indicate that you would like to keep this open!

**Comment by juscuzryancan** [comment]
@raclim I would like to keep this open and I will provide updates soon

**Comment by griffinwalker** [comment]
yes please

**Comment by juscuzryancan** [comment]
Reopening in a separate PR

**Pull request closed without merging**